aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1479 @@
+    fun pending_withdrawal_exists(pool: &DelegationPool, delegator_address: address): (bool, u64) {
+        if (table::contains(&pool.pending_withdrawal, delegator_address)) {
+            (true, *table::borrow(&pool.pending_withdrawal, delegator_address))
+        } else {
+            (false, 0)
+        }
+    }
+
+    fun charge_add_stake_fee(pool: &DelegationPool, coins_amount: u64): u64 {
+        let pool_address = get_pool_address(pool);
+        // if the underlying stake pool earns rewards this epoch, charge delegator
+        // the maximum amount it would earn from new added stake in `pending_active` state
+        if (stake::is_current_epoch_validator(pool_address)) {
+            let active_shares = &pool.active_shares;
+            let (active, _, pending_active, _) = stake::get_stake(pool_address);
+
+            let (rewards_rate, rewards_rate_denominator) = staking_config::get_reward_rate(&staking_config::get());
+            let max_epoch_active_rewards = if (rewards_rate_denominator > 0) {
+                pool_u64::multiply_then_divide(active_shares, active, rewards_rate, rewards_rate_denominator)
+            } else {
+                0
+            };
+
+            // 1. calculate shares received if buying in active pool with its pending epoch rewards added
+            // 2. calculate coins required to buy this amount of shares in current active pool
+            // (((coins_amount * total_shares) / (total_coins + max_epoch_active_rewards)) * total_coins) / total_shares
+            pool_u64::multiply_then_divide(
+                active_shares,
+                coins_amount,
+                // synchronized total_coins on `active_shares`
+                active + pending_active,
+                active + pending_active + max_epoch_active_rewards
+            )
+        } else {
+            coins_amount
+        }
+    }
+
+    /// Buy shares into active pool on behalf of delegator `shareholder` who
+    /// delegated `coins_amount` stake or reactivated(redeemed) it from pending-inactive pool.
+    fun buy_in_active_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+    ): u64 {
+        if (coins_amount == 0) return 0;
+
+        pool_u64::buy_in(&mut pool.active_shares, shareholder, coins_amount)
+    }
+
+    /// Buy shares into pending-inactive pool on behalf of delegator `shareholder` who 
+    /// redeemed `coins_amount` from active pool to schedule it for unlocking.
+    /// If there is a pending withdrawal from a past epoch, fail the operation.
+    fun buy_in_inactive_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+    ): u64 {
+        if (coins_amount == 0) return 0;
+
+        // save lockup epoch for new pending withdrawal if no existing previous one
+        let current_lockup_epoch = pool.current_lockup_epoch;
+        assert!(*table::borrow_mut_with_default(
```

Comment:
Are you using borrow_mut_with_default because there's no borrow_with_default?

Reply:
it is a concise way of trying to set the pending withdrawal for delegator: if there is one already present at past lockup < `current_lockup_epoch` then the assert fails this unlock operation, otherwise set the lockup cycle for this new pending withdrawal and move forward.